src/poetry/publishing/uploader.py

```diff
@@ -50,6 +50,30 @@ def __init__(self, error: ConnectionError | HTTPError | str) -> None:
         super().__init__(message)
 
 
+def _ignore_existing(response: requests.Response, skip_existing: bool) -> bool:
+    # based on https://github.com/pypa/twine/blob/62cec5b77037345b4fe1a85bbc6b104440799496/twine/commands/upload.py#L32  # noqa: E501
+    if not skip_existing:
+        return False
+
+    status = response.status_code
+    reason = getattr(response, "reason", "").lower()
+    text = getattr(response, "text", "").lower()
+
+    # NOTE(sigmavirus24): PyPI presently returns a 400 status code with the
+    # error message in the reason attribute. Other implementations return a
+    # 403 or 409 status code.
+    return (
+        # pypiserver (https://pypi.org/project/pypiserver)
+        status == 409
+        # PyPI / TestPyPI
+        or (status == 400 and "already exist" in reason)
+        # Nexus Repository OSS (https://www.sonatype.com/nexus-repository-oss)
+        or (status == 400 and "updating asset" in reason)
+        # Artifactory (https://jfrog.com/artifactory/)
+        or (status == 403 and "overwrite artifact" in text)
+    )
```

Comment:
```suggestion
    status_codes = [
        409,  # pypiserver (https://pypi.org/project/pypiserver)
    ]
    status_keywords = {
        400: [
            # PyPI / TestPyPI / GCP Artifact Registry
            "already exist",
            # Nexus Repository OSS (https://www.sonatype.com/nexus-repository-oss)
            "updating asset",
            # Gitlab Enterprise Edition (https://about.gitlab.com)
            "already been taken",
        ],
        403: [
            # Artifactory (https://jfrog.com/artifactory/)
            "overwrite artifact",
        ],
    }
    return status in status_codes or (
        status in status_keywords
        and any(
            keyword in message
            for keyword in status_keywords[status]
            for message in [reason, text]
        )
    )
```

It seems there have been updates to this function we should adopt (https://github.com/pypa/twine/blob/main/twine/commands/upload.py). Maybe, we could reformulate and even generalize this a bit (just a rough proposal). I'm aware it is not 100% equivalent to `twine` anymore, but I think it is not worse.

Reply:
i don't mind straying from the twine code, though i do find the original or something closer to it easier to read than pulling things out into `status_keywords` etc. will look at the updated twine and propose something